File: library/src/main/java/com/gmail/fuskerr63/android/library/presenter/contact/ContactDetailsPresenter.java
package com.gmail.fuskerr63.android.library.presenter.contact;

import android.util.Log;

import com.gmail.fuskerr63.java.entity.Contact;
import com.gmail.fuskerr63.java.entity.ContactInfo;
import com.gmail.fuskerr63.java.interactor.DatabaseInteractor;
import com.gmail.fuskerr63.android.library.view.ContactDetailsView;
import com.gmail.fuskerr63.java.interactor.ContactInteractor;
import com.gmail.fuskerr63.java.interactor.NotificationInteractor;
import com.gmail.fuskerr63.library.BuildConfig;

import java.util.Objects;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.Nullable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import moxy.MvpPresenter;

public class ContactDetailsPresenter extends MvpPresenter<ContactDetailsView> {
    @Nullable
    private final ContactInteractor contactInteractor;
    @Nullable
    private final DatabaseInteractor databaseInteractor;
    @Nullable
    private final NotificationInteractor notificationInteractor;

    private final CompositeDisposable disposable = new CompositeDisposable();

    @Inject
    public ContactDetailsPresenter(
            @Nullable ContactInteractor contactInteractor,
            @Nullable DatabaseInteractor databaseInteractor,
            @Nullable NotificationInteractor notificationInteractor) {
        this.contactInteractor = contactInteractor;
        this.databaseInteractor = databaseInteractor;
        this.notificationInteractor = notificationInteractor;
    }

    public void showDetails(int id) {
        if (contactInteractor != null) {
            disposable.add(contactInteractor.getContactById(id)
                    .subscribeOn(Schedulers.io())
                    .flatMap(contact -> databaseInteractor.getUserByContactId(contact.getId())
                            .map(user -> {
                                ContactInfo contactInfo = new ContactInfo(
                                        contact.getName(),
                                        contact.getNumber(),
                                        contact.getNumber2(),
                                        contact.getEmail(),
                                        contact.getEmail2()
                                );
                                return new Contact(
                                        contact.getId(),
                                        contact.getImage(),
                                        contactInfo,
                                        contact.getBirthday(),
                                        user.getAddress()
                                );
                            }))
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe(d -> getViewState().loadingStatus(true))
                    .doFinally(() -> getViewState().loadingStatus(false))
                    .subscribe(
                            contact -> getViewState().updateDetails(contact),
                            error -> {
                                if (BuildConfig.DEBUG) {
                                    Log.d("TAG", Objects.requireNonNull(error.getMessage()));
                                }
                            }
                    ));
        }
    }

    public void onClickBirthday(
            @Nullable Contact contact,
            @Nullable String notificationCancel,
            @Nullable String notificationSend) {
        if (notificationInteractor != null
                && notificationInteractor.toggleNotificationForContact(contact).isAlarmUp()) {
            getViewState().setTextButton(notificationCancel);
        } else {
            getViewState().setTextButton(notificationSend);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposable.dispose();
    }
}

File: library/src/main/java/com/gmail/fuskerr63/android/library/presenter/contact/ContactDetailsPresenter.kt
package com.gmail.fuskerr63.android.library.presenter.contact

import android.util.Log
import com.gmail.fuskerr63.android.library.view.ContactDetailsView
import com.gmail.fuskerr63.java.entity.Contact
import com.gmail.fuskerr63.java.entity.ContactLocation
import com.gmail.fuskerr63.java.interactor.ContactInteractor
import com.gmail.fuskerr63.java.interactor.DatabaseInteractor
import com.gmail.fuskerr63.java.interactor.NotificationInteractor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flatMapMerge
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import moxy.MvpPresenter
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

class ContactDetailsPresenter @Inject constructor(
    private val contactInteractor: ContactInteractor,
    private val databaseModel: DatabaseInteractor,
    private val notificationInteractor: NotificationInteractor
) : MvpPresenter<ContactDetailsView?>(), CoroutineScope {
    override val coroutineContext: CoroutineContext = SupervisorJob() + Dispatchers.Main

    @FlowPreview
    fun showDetails(id: Int) {
        if (id != -1) {
            @Suppress("TooGenericExceptionCaught")
            try {
                launch {
                    contactInteractor.getContactById(id)
                        .flatMapMerge { contact: Contact ->
                            databaseModel.getFlowUserById(contact.id)
                                .map { value: ContactLocation? ->
                                    createContactWithLocation(contact, value)
                                }
                        }
                        .flowOn(Dispatchers.IO)
                        .collect { contact: Contact? ->
                            viewState?.updateDetails(contact)
                            viewState?.setTextButton(notificationInteractor.getNotificationStatusForContact(contact))
                        }
                }
            } catch (error: Exception) {
                Log.d("TAG", error.message ?: "")
            }
        }
    }

    private fun createContactWithLocation(contact: Contact, location: ContactLocation?) =
        with(contact) {
            Contact(
                id,
                image,
                contactInfo,
                birthday,
                location?.address ?: ""
            )
        }
    fun onClickBirthday(contact: Contact?) {
        viewState?.setTextButton(notificationInteractor.toggleNotificationForContact(contact))
    }

    override fun onDestroy() {
        cancel()
        super.onDestroy()
    }
}
